Tidy Chef cookbooks

bleggett opened the pull request:
(I see a new release branch was cut, if this should be targeted against that let me know and I'll retarget)

For some time we have had two different chef cookbooks, `jenkins` and `blankJenkins`. We only ever actually invoke the `configureBlankJenkins` recipe from the `blankJenkins` cookbook, but that recipe calls back to support scripts in the `jenkins` cookbook, which is messy and confusing.

Apparently at one time these two cookbooks were supposed to be consolidated, but the consolidation was never finished. 

- I've moved the relevant `configureBlankJenkins` recipe from `blankJenkins` into `jenkins`, renamed it to `configureJenkins`, and deleted the `blankJenkins` cookbook as the remainder of it was unused cruft.

- I've removed confusing/redundant cookbook copy steps from `chefProvisioner.tf`.

- I've bumped the `terraform` and `chefdk` versions to 0.11.7 and 1.6.11 respectively.

~~(The chef provisioning steps succeed for me with Scenario 3, but I haven't been able to test a full E2E install as I'm getting unrelated DynamoDB rate limit errors in my AWS account which I need to sort out.)~~

EDIT: Scenario 3 passes cleanly.

bleggett commented:
Rebased PR on 1.5.1

bleggett commented on installscripts/jazz-terraform-unix-noinstances/scripts/modifyCodebase.sh:
```diff
@@ -21,67 +21,67 @@ timestamp=`date --utc +%FT%T`
+
+    if [ $element == "jazz_email" ] || [ $element == "jazz_usermanagement" ] ; then
+        aws dynamodb put-item --table-name $tablename --item '{
+      "SERVICE_ID":{"S":"'$uuid'"},
```
@svidhani It looks like in this script we are setting up Jazz core services in this script using both node6 and node4 provider runtimes.

Do those services actually need different node runtime versions, or can set all of them to run against node6?

bleggett replied:
Alright, I attempted to parameterize and also bumped to Node 8.10, which is the most recent version Lambda supports.

suryajak replied:
We have a mix of old and new services which are based on ES5 and ES6, so we will need to continue to build and run them with the appropriate node versions. We have a plan to migrate to ES8 next quarter.

svidhani replied:
yes we need both as some services are using node4 and some are on node6. eventually we want to move to 6 first. We are not on 8 yet.

bleggett replied:
Alright, in that case reverted the changes to the node runtime flags.

Suggest that when we get them all on the same runtime version, we parameterize the node version string then.

suryajak commented on installscripts/jazz-terraform-unix-noinstances/scripts/DeleteStackPlatformServices.py:
```diff
@@ -41,20 +41,21 @@ def deleteCloudFormationService(service_name):
 
 for pservice in platformServices:
-    deleteCloudFormationService(stackName + pservice + '-prod')
+    deleteCloudFormationService(stackName + pservice)
```
We need `-prod`, bad merge probably?

bleggett replied:
Dangit, you're right, fixed this and rechecked every conflict from that merge.

rajeevr2715 commented on installscripts/jazz-terraform-unix-noinstances/chefProvisioner.tf:
```diff
@@ -90,59 +90,9 @@ resource "null_resource" "chef_provision_jenkins_server" {
     ]
   }
 
-  provisioner "file" {
```
@bleggett Got an error while deploying. 

> HTTP Request Returned 412 Precondition Failed: No such cookbook: git
> Error Resolving Cookbooks for Run List

bleggett replied:
What scenario? With Scenario 3 I get the correct result.

rajeevr2715 replied:
@bleggett I think this is already discussed her - https://github.com/tmobile/jazz-installer/pull/136
Please take a look.

bleggett replied:
Right, I remember this. The way to fix this once and for all is to stop copying things to $HOME.

Because $HOME changes depending on the scenario and whether we run chef-client as sudo, and if we use home we can't avoid hardcoding a user-specific path into the chef cookbooks, and we definitely want to avoid that.

Going to test out a change to copy the cookbooks to `/tmp`, that should work in all cases and should be pretty agnostic for all POSIX systems.

suryajak commented on installscripts/cookbooks/aws/recipes/default.rb:
```diff
@@ -18,6 +18,5 @@
 end
 
 execute 'installaws-apigateway-importer' do
```
we got rid of apigateway-importer in our v1.4.6 branch, perhaps this is a left over, can we remove it?

bleggett commented:
Rebased on 1.5.2

Background: In order to address the longstanding problem @rajeevr2715 bought up a few days back, went over the chef cookbooks.

Turns out most of the recipies were unused, and the one we _did_ use was a giant IF statement with two branches, where the only difference between the branches was whether we were rooting paths in `/home/USER` or `/root`.

Made the prefix a variable, and deleted all the duplicated chef code. Now we have _one_ recipe, `configurejenkins`, which runs for all scenarios. Also updated a bunch of the helper scripts that were hardcoding/copypasting stuff they didn't need to hardcode.

The Chef cookbooks still need some work, and I plan to delete the `pip` install logic in Terraform and all the aws/node Chef cookbooks and replace them with off-the-shelf cookbooks from the Chef Supermarket which will be installed at runtime.

But I plan to do that in a later PR, this one is already giant enough.

@rajeevr2715 This one passes Scenario 3 for me, please test with 1. I deleted a bunch of stuff that had no references, but it's always possible/likely I broke something in a scenario I can't test.

bleggett commented:
@suryajak With an AWS Large instance and a Jenkins container (scenario 3), I had no problems with deleting the 45-sec per loop sleep. I feel like unless people are using servers that are too small we should not need to wait several minutes per install.

But I went ahead and put the sleep back inside the loop, and changed it from 45 secs to 10 secs anyway. If that causes problems, will bump it back to 45

bleggett commented:
@rajeevr2715 Yep, and if you look at `jazz_jenkins.py` you'll see that I changed that check to do what the Chef script does: Download the `jenkins-cli.jar` file from the Jenkins server itself at runtime, rather than relying on a checked-in binary. So we don't need it anymore.

This is the recommended approach anyway, you're supposed to use the jar file that matches your server version.

bleggett commented:
@rajeevr2715 Sorry, in another branch I had made more aggressive changes, altering `jazz-jenkins.py` to fetch the jar file from the Jenkins server so we don't need the checked in version, and didn't port it to this branch. I will port that fix over.

This is the recommended approach in the Jenkins docs anyway, you're supposed to use the jar file that matches your server version. If we rely on the checked-in version there's no guarantee people won't have errors because the jar file won't match their server version.

bleggett commented:
I also see someone added a `sonar.sh` which is largely identical to the scripts being run in the chef cookbooks, it should be run alongside its brothers and sisters in the cookbook rather than via Terraform directly, but I will fix that after this PR is merged and the Chef cookbooks are more managable.

bleggett commented:
@rajeevr2715 Both jarfile fix and missed `sonar.sh` change pushed.

Note that there are a lot of currently-unused scripts in `/cookbooks/jenkins/files/` that I have left alone in this PR. I have another PR dependent on this one that will remove the unused scripts which I will submit once this one is merged and I work out some kinks.

suryajak commented:
@rajeevr2715 : Can you please CR as well as validate this PR against scenario 1 and 3?

rajeevr2715 commented:
@bleggett got one error, 

> Error: resource 'null_resource.chef_provision_jenkins_server' provisioner local-exec (#8): unknown variable referenced: 'cookbooksDir'; define it with a 'variable' block

bleggett commented:
@rajeevr2715 Fixed. Re-synched with recent changes made to 1.5.2 and didn't re-test, so missed that one.

rajeevr2715 commented:
@bleggett For scenario1, The bitbucket server configured for both Jazz_Core_Services and Jazz_User_Services- **SCMURL/scm** was not needed.

Also, got error while running buildpack jenkins jobs - 

> /var/lib/jenkins/workspace/build_pack_api@tmp/durable-*****/script.sh: line 2: aws: command not found


the below lines were needed, I guess.

> -link '/usr/bin/aws' do
> -  to '/usr/local/bin/aws'
> -  owner 'root'
> -  group 'root'
> -  mode '0777'
> -end

bleggett commented on installscripts/jazz-terraform-unix-noinstances/chefProvisioner.tf:
```diff
@@ -44,13 +44,13 @@ resource "null_resource" "chef_provision_jenkins_server" {
   }
 
   provisioner "local-exec" {
     command = "sed -i 's|<password>sonarpasswd</password>|<password>${lookup(var.codeqmap, "sonar_passwd")}</password>|g' ${var.cookbooksDir}/jenkins/files/credentials/sonar.sh"
```
Dangit, missed that, fixed.

bleggett commented:
```
@bleggett For scenario1, The bitbucket server configured for both Jazz_Core_Services and Jazz_User_Services- SCMURL/scm was not needed.
```
You mean parts of the configuration pushed in `job_platform_api_services.sh` are unnecessary in scenario 1 (because there is no bitbucket in that case)? 

Since that's an existing issue that isn't currently causing any problems, I will look at correcting that in the next PR. I mainly want to get the cookbook into a non-crappy state first.

```
Also, got error while running buildpack jenkins jobs -

    /var/lib/jenkins/workspace/build_pack_api@tmp/durable-*****/script.sh: line 2: aws: command not found

the below lines were needed, I guess.

    -link '/usr/bin/aws' do

        to '/usr/local/bin/aws'
        owner 'root'
        group 'root'
        mode '0777'
        -end
```

That should not be the case unless something in `script.sh` is incorrect. Is that script in `jazz` or `jazz-installer`?

bleggett commented:
@rajeevr2715 I think the problem with the `aws` binary is actually here in `jazz-installer/installscripts/cookbooks/aws/recipes/default.rb`:

```
execute 'installaws' do
  command 'sudo ./awscli-bundle/install -i /usr/local/aws -b /usr/local/bin/aws'
  cwd '/tmp'
  ignore_failure true
end
```

My guess is for scenario 1 that fails, because `centos` user doesn't have perms to write to `/usr/local/bin`, and since someone decided adding the `ignore_failure` flag was a good idea, we never noticed that the binary install failed.

So that old `link` step was added as a hacky workaround.

Pushed a change running the above with `sudo` and removing the `ignore_failure` flag.

bleggett commented:
@suryajak @rajeevr2715 Should I retarget this to 1.6?

suryajak commented:
The current target branch sounds ok to me. Our release tag for branch v1.5.2 will be 1.7.

rajeevr2715 commented:
@bleggett I have tested and got the same issues again. In order to make it running, I have to add an environment varibale inside the jenkins configuration like this,
>Path
> /usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/apache-maven-3.5.2/bin:/home/centos/.local/bin:/home/centos/bin:/opt/apache-maven-3.5.2/bin:/home/centos/.local/bin:/home/centos/bin

![bitbucket_server_with_scm_appended](https://user-images.githubusercontent.com/36688599/40932558-5c37a7b2-684c-11e8-8250-fe7275dafb85.png)
![env_jenkins](https://user-images.githubusercontent.com/36688599/40932559-5c78d156-684c-11e8-8c6f-bf661b314488.png)

rajeevr2715 commented:
@bleggett   I got the same issues again.

In order to make it work, I have to add Path environment variable inside jenkins configuration

![env_jenkins](https://user-images.githubusercontent.com/36688599/40932749-ffb127a6-684c-11e8-9aa8-d0863c365e2c.png)
/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/apache-maven-3.5.2/bin:/home/centos/.local/bin:/home/centos/bin:/opt/apache-maven-3.5.2/bin:/home/centos/.local/bin:/home/centos/bin

Also, I have to remove "/scm" from the Bitbucket Server UR parameter inside  - Projects -  Bitbucket Team/Project

bleggett commented:
@rajeevr2715 Can you log into that Jenkins box as the `centos` user and paste the results of `echo $PATH` here?


In a normal Linux system `/usr/bin/` MUST be in PATH, you should not need to specify it unless the box is jacked up. And most that I've encountered also have `/usr/local/bin` in PATH.

It's possible that whatever variant of Linux your Jenkins test box is in does not have `/usr/local/bin` in PATH, so let me change that cookbook to at least put the binary in `/usr/bin`. If it's not found there, then that box's PATH variable is broken (or Jenkins does something odd with it).

rajeevr2715 commented:
@bleggett Actually the jenkins job failed because it could not identify the aws command. Most probably because of the path. I have added the path to jenkins manually to test whether it will work after making the path variable as en environment variable. After that the jenkins never failed on aws command not found error. Since these steps are not actually needed before this change and the jenkins never failed with aws not found error earlier. So we need to find what missed actually.

Regarding the Bitubucket SCM path, Will look into it why it was failing. Might be the issue was already there and we could not identify since our test jenkins server has an already copied cookbooks folder.

bleggett commented:
@rajeevr2715 
So you're saying that after changing the Jenkins path setting to 

```
$PATH:/opt/apache-maven-3.5.2/bin:/home/centos/.local/bin:/home/centos/bin:/opt/apache-maven-3.5.2/bin:/home/centos/.local/bin:/home/centos/bin
```

it still did not work?

rajeevr2715 commented:
@bleggett All those failed, I had to give the path inside the jenkins configuration as 

> $PATH:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/apache-maven-3.5.2/bin:/home/centos/.local/bin:/home/centos/bin

then removed scm from the bitbucket server inside Jazz Core services and Jazz User services and ran all the jenkins job. The stack is up now for scenario 1.

@suryajak So we had an extra configuration ie, add environment path variable in the scenario 1 existing jenkins. Also need to fix the bitbucket scm url inside jenkins for  Jazz Core services and Jazz User services

bleggett commented:
@rajeevr2715 If this one 

```
$PATH:/opt/apache-maven-3.5.2/bin:/home/centos/.local/bin:/home/centos/bin:/opt/apache-maven-3.5.2/bin:/home/centos/.local/bin:/home/centos/bin
```
failed that still makes me think there is a problem with the default $PATH on that box, or for the `jenkins` user.

Or perhaps Jenkins does something nonstandard by default, like clearing the default $PATH for the `jenkins` user.

Either way, the way you've got it working now is better. Chef cookbooks (and all other provisioning tools) need to be assured of a somewhat-standard Linux baseline environment to be useful.

bleggett commented:
@rajeevr2715 Thanks for reviewing/testing! I know scenario 1 is a pain.

bleggett commented:
Resynced with 1.5.2 due to drift.
